libavcodec/jpeg2000dec.c

```diff
@@ -1067,100 +1072,343 @@ static int jpeg2000_decode_packet(Jpeg2000DecoderContext *s, Jpeg2000Tile *tile,
+                            av_log(s->avctx, AV_LOG_ERROR,
+                                   "nonzerobits %d invalid or unsupported\n", v);
+                            return AVERROR_INVALIDDATA;
+                        }
+                        cblk->nonzerobits = v;
+                    }
+                    cblk->zbp = zbp;
+
+                    cblk->lblock = 3;
                 }
-                cblk->zbp = zbp;
-                cblk->nonzerobits = v;
-            }
-            if ((newpasses = getnpasses(s)) < 0)
-                return newpasses;
-            av_assert2(newpasses > 0);
-            if (cblk->npasses + newpasses >= JPEG2000_MAX_PASSES) {
-                avpriv_request_sample(s->avctx, "Too many passes");
-                return AVERROR_PATCHWELCOME;
-            }
-            if ((llen = getlblockinc(s)) < 0)
-                return llen;
-            if (cblk->lblock + llen + av_log2(newpasses) > 16) {
-                avpriv_request_sample(s->avctx,
-                                      "Block with length beyond 16 bits");
-                return AVERROR_PATCHWELCOME;
+            } else {
+                incl = get_bits(s, 1);
             }
 
-            cblk->lblock += llen;
-
-            cblk->nb_lengthinc = 0;
-            cblk->nb_terminationsinc = 0;
-            av_free(cblk->lengthinc);
-            cblk->lengthinc = av_calloc(newpasses, sizeof(*cblk->lengthinc));
-            if (!cblk->lengthinc)
-                return AVERROR(ENOMEM);
-            tmp = av_realloc_array(cblk->data_start, cblk->nb_terminations + newpasses + 1, sizeof(*cblk->data_start));
-            if (!tmp)
-                return AVERROR(ENOMEM);
-            cblk->data_start = tmp;
-            do {
-                int newpasses1 = 0;
-
-                while (newpasses1 < newpasses) {
-                    newpasses1 ++;
-                    if (needs_termination(codsty->cblk_style, cblk->npasses + newpasses1 - 1)) {
-                        cblk->nb_terminationsinc ++;
-                        break;
+            if (incl) {
+                int nb_segments = 0;
+                uint16_t *segment_length = NULL;
+                uint8_t bypass_term_threshold = 0;
+                uint8_t bits_to_read = 0;
+                uint8_t pass_index = cblk->npasses;
+                uint32_t segment_bytes = 0;
+                int32_t segment_passes = 0;
+                uint8_t next_segment_passes = 0;
+                int32_t href_passes, pass_bound;
+                uint8_t primary_passes, secondary_passes;
+                uint32_t primary_bytes, secondary_bytes;
+                uint32_t fast_skip_bytes = 0;
+                int empty_set;
+                uint32_t tmp_length = 0;
+
+                if ((newpasses = getnpasses(s)) < 0)
+                    return newpasses;
+                if(newpasses <= 0)
+                    av_log(s->avctx, AV_LOG_ERROR, "Invalid number of passes is found in the packet header.\n");
```

Comment:
Should we exit here?

Reply:
`newpasses < 0` never happens at line 1123 because the negative value of `newpasses` should be checked in line 1121. Actually, the condition in line 1121 can be `<= 0` because non-zero `incl` means `newpasses` is at least one.

I have checked with the condition, `<=0`, in line 1121 for all the test cases in 15444-4. All results are `passed` for the conformance thresholds.

In conclusion, lines 1123-1124 can safely be removed without any problem.